### PR TITLE
[FlexAttention] Support training bias for eager (#136910)

### PR DESCRIPTION
**Pull request by drisspg**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137526

This PR is Part 2 of the implementation started in https://github.com/pytorch/pytorch/pull/136910, rolled in the updates from https://github.com/pytorch/pytorch/pull/137451. Original was reverted due to calls to #@torch.libary at `import torch` time, so added a call to register at first call to `ModIndex`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec

### CODE REVIEW

**Comment by vadimkantorov** [comment]
@drisspg I wonder if FlexAttention could be added as a backend for existing SDPA method, maybe then it would be simpler for existing SDPA workloads to switch to it - for benchmarking flex attention in the basic setup (without score mods etc)...

File: torch/_higher_order_ops/flex_attention.py [hunk 1]
```diff
@@ -69,6 +69,79 @@ def _permute_strides(out: torch.Tensor, query_strides: Tuple[int, ...]) -> torch
...
+        return out, None
+
+
+class ModIndex(torch.autograd.Function):
+    _is_registered = False
+    generate_vmap_rule = True
+
+    @staticmethod
+    def forward(x: Tensor, indices: List[Tensor]) -> Tensor:
+        if not ModIndex._is_registered:
```
Comment (drisspg) [inline torch/_higher_order_ops/flex_attention.py:118]: Do I need to add any locks around this to avoid any threading issues?

**Comment by drisspg** [comment]
@vadimkantorov  there has been some talk abou tthis and it might make sense if sdpa is called under compile, feel free to create a separate issue for this

File: test/inductor/test_flex_attention.py [hunk 2]
```diff
@@ -989,14 +989,15 @@ def composed_score_mod(score, b, h, m, n):
         self.run_test(composed_score_mod, dtype)
 
     @supported_platform
+    # @expectedFailure  # TODO: Remove this after supporting compiled flex attention with training bias
```
Comment (drisspg) [inline test/inductor/test_flex_attention.py:992]: I dont really trust this...

Comment (Chillee) [reply test/inductor/test_flex_attention.py:992]: You don't believe in expectedFailure?

**Review by Chillee — APPROVED** [review]

File: test/inductor/test_flex_attention.py (hunk 2 above)
Comment (drisspg) [reply test/inductor/test_flex_attention.py]: This is why I didnt trust this:

https://github.com/pytorch/pytorch/pull/137526/files#diff-e3963412cc249e81fecfcf8774f5428b2b5e837ff3633ae13d8b7886ab5bc3b9R2074

**Review by zou3519 — APPROVED** [review]

File: torch/_higher_order_ops/flex_attention.py [hunk 3]
```diff
@@ -94,17 +91,23 @@ def _(
...
-            for i, idx in enumerate(indices):
-                # TODO: Don't use is_batchedtensor API, use in_dims instead.
-                if torch._C._functorch.is_batchedtensor(idx):
-                    indices[i] = idx.unsqueeze(-1)
+        indices_indims = indims[1]
+        expanded_indices = []
+        for idx, idx_indim in zip(indices, indices_indims):
+            # The index is not a being batched, we should unsqueeze and expand to val
+            if idx_indim is None:
+                expanded_indices.append(idx.unsqueeze(0).expand(value.shape))
```
Comment (vadimkantorov) [inline torch/_higher_order_ops/flex_attention.py:101]: could simply `.append(idx.expand_as(value))` be used instead?

also maybe explicit `unsqueeze(0)` could be omitted? as broadcasting should already do it

Comment (drisspg) [reply torch/_higher_order_ops/flex_attention.py:101]: Good tip, thanks

**Comment by drisspg** [comment]
@mlazos verifeid that if I turn of _compile=False in the failing tst for the create_block_mask function error goes away, any ideas what might be happening here?

**Comment by drisspg** [comment]
@mlazos w this modification to https://github.com/pytorch/pytorch/pull/137809 I get the tests to pass locally
```Shell
❯ g dft
torch/_dynamo/variables/misc.py --- Python
712 712         ctx = AutogradFunctionContextVariable.create(tx, args, kwargs)
713 713         args = [ctx, *args]
714 714         if isinstance(fn, types.FunctionType):
... 715             sig = inspect.signature(fn)
... 716             if len(args) - 1  == len(sig._parameters):
... 717                 args = args[1:]  # Don't use context
715 718             return variables.UserFunctionVariable(fn, source=source).call_function(
716 719                 tx, args, kwargs
717 720             )

```

**Pull request closed without merging**